driver/parallel.go

```diff
@@ -0,0 +1,222 @@
+package driver
+
+import (
+	"sync"
+
+	"github.com/brimsec/zq/ast"
+	"github.com/brimsec/zq/expr"
+	"github.com/brimsec/zq/filter"
+	"github.com/brimsec/zq/proc"
+	"github.com/brimsec/zq/scanner"
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zng"
+)
+
+type parallelHead struct {
+	proc.Base
+	once sync.Once
+	pg   *parallelGroup
+
+	mu sync.Mutex // protects below
+	sc ScannerCloser
+}
+
+func (ph *parallelHead) closeOnDone() {
+	select {
+	case <-ph.Context.Done():
+		ph.mu.Lock()
+		defer ph.mu.Unlock()
+		if ph.sc != nil {
+			ph.sc.Close()
+			ph.sc = nil
+		}
+	}
+}
+
+func (ph *parallelHead) Pull() (zbuf.Batch, error) {
+	// Trigger the parallel group to read from the multisource.
+	ph.pg.once.Do(func() {
+		go ph.pg.run()
+	})
+	// Ensure open scanners are closed when flowgraph execution stops.
+	ph.once.Do(func() {
+		go ph.closeOnDone()
+	})
+
+	ph.mu.Lock()
+	defer ph.mu.Unlock()
+
+	for {
+		if ph.sc == nil {
+			sc, err := ph.pg.nextSource()
+			if sc == nil || err != nil {
+				return nil, err
+			}
+			ph.sc = sc
+		}
+		batch, err := ph.sc.Pull()
+		if err != nil {
+			return nil, err
+		}
+		if batch == nil {
+			if err := ph.sc.Close(); err != nil {
+				return nil, err
+			}
+			ph.pg.doneSource(ph.sc)
+			ph.sc = nil
+			continue
+		}
+		return batch, err
+	}
+}
+
+type parallelGroup struct {
+	filter     SourceFilter
+	msrc       MultiSource
+	once       sync.Once
+	pctx       *proc.Context
+	sourceChan chan SourceOpener
+	sourceErr  error
+
+	mu       sync.Mutex // protects below
+	stats    scanner.ScannerStats
+	scanners []scanner.Scanner
+}
+
+func (pg *parallelGroup) nextSource() (ScannerCloser, error) {
+	for {
+		select {
+		case opener, ok := <-pg.sourceChan:
+			if !ok {
+				return nil, pg.sourceErr
+			}
+			sc, err := opener()
+			if err != nil {
+				return nil, err
+			}
+			if sc == nil {
+				continue
+			}
+			pg.mu.Lock()
+			pg.scanners = append(pg.scanners, sc)
+			pg.mu.Unlock()
+			return sc, nil
+		case <-pg.pctx.Done():
+			return nil, pg.pctx.Err()
+		}
+	}
+}
+
+func (pg *parallelGroup) doneSource(sc ScannerCloser) {
+	pg.mu.Lock()
+	defer pg.mu.Unlock()
+	pg.stats.Accumulate(sc.Stats())
+	for i := range pg.scanners {
+		if pg.scanners[i] == sc {
+			pg.scanners[i] = nil
+			pg.scanners = append(pg.scanners[:i], pg.scanners[i+1:]...)
+			break
+		}
+	}
+}
+
+func (pg *parallelGroup) Stats() *scanner.ScannerStats {
+	pg.mu.Lock()
+	defer pg.mu.Unlock()
+	s := pg.stats
+	for _, sc := range pg.scanners {
+		s.Accumulate(sc.Stats())
+	}
+	return &s
+}
+
+func (pg *parallelGroup) run() {
+	defer close(pg.sourceChan)
+	pg.sourceErr = pg.msrc.SendSources(pg.pctx, pg.pctx.TypeContext, pg.filter, pg.sourceChan)
```

Comment:
Nit: No need for defer.
```suggestion
	pg.sourceErr = pg.msrc.SendSources(pg.pctx, pg.pctx.TypeContext, pg.filter, pg.sourceChan)
	close(pg.sourceChan)
```